File: backend/app/schemas/geometry.py
from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, Field


class BoundingBox(BaseModel):
    min: list[float]
    max: list[float]
    size: list[float]


class GeometryTransformIn(BaseModel):
    """Preparation step params. Re-applying always starts from the original
    upload (see Geometry.transformations docstring) -- this is a full
    description of the desired state, not a delta."""
    unit: Literal["mm", "cm", "m", "in"] = "mm"
    scale: float = Field(1.0, gt=0, le=1_000_000)
    rotate_x: float = Field(0.0, ge=-360, le=360)
    rotate_y: float = Field(0.0, ge=-360, le=360)
    rotate_z: float = Field(0.0, ge=-360, le=360)
    up_axis: Literal["x", "y", "z"] = "z"
    center_xy: bool = False
    ground_align: bool = False
    invert_normals: bool = False
    simplify_target_faces: Optional[int] = Field(None, gt=3, le=5_000_000)


class VoxelizeIn(BaseModel):
    geometry_id: str
    resolution: int = Field(64, ge=8, le=1024)
    fill_interior: bool = True


class GeometryPreviewPayload(BaseModel):
    vertices: list[list[float]]
    triangles: list[list[int]]
    truncated: bool


class GeometryOut(BaseModel):
    id: str
    project_id: Optional[str] = None
    original_filename: str
    format: str
    status: str
    dimension: str
    units: str
    vertex_count: Optional[int] = None
    face_count: Optional[int] = None
    point_count: Optional[int] = None
    bounds: Optional[BoundingBox] = None
    center: Optional[list[float]] = None
    watertight: Optional[bool] = None
    normals_consistent: Optional[bool] = None
    transformations: dict
    grid_path: Optional[str] = None
    grid_shape: Optional[list[int]] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class GeometryUploadOut(GeometryOut):
    size_mb: float
    preview: Optional[GeometryPreviewPayload] = None


class GeometryPrepareOut(GeometryOut):
    preview: Optional[GeometryPreviewPayload] = None
    notes: list[str] = []


class VoxelizeOut(BaseModel):
    geometry_id: str
    grid_id: str
    grid_path: str
    grid_shape: list[int]
    total_cells: int
    solid_cells: int
    fluid_cells: int

File: frontend/src/types/geometry.ts
export type GeometryDimension = '2D' | '3D';

export type GeometryStatus = 'uploaded' | 'analyzed' | 'prepared' | 'voxelized' | 'error';

export interface BoundingBox {
  min: number[];
  max: number[];
  size: number[];
}

export interface MeshPreviewPayload {
  vertices: number[][];
  triangles: number[][];
  truncated: boolean;
}

export interface GeometryRecord {
  id: string;
  project_id: string | null;
  original_filename: string;
  format: string;
  status: GeometryStatus;
  dimension: GeometryDimension;
  units: string;
  vertex_count: number | null;
  face_count: number | null;
  point_count: number | null;
  bounds: BoundingBox | null;
  center: number[] | null;
  watertight: boolean | null;
  normals_consistent: boolean | null;
  transformations: GeometryTransform;
  grid_path: string | null;
  grid_shape: number[] | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface GeometryUploadResult extends GeometryRecord {
  size_mb: number;
  preview: MeshPreviewPayload | null;
}

export interface GeometryPrepareResult extends GeometryRecord {
  preview: MeshPreviewPayload | null;
  notes: string[];
}

export type LengthUnit = 'mm' | 'cm' | 'm' | 'in';
export type UpAxis = 'x' | 'y' | 'z';

export interface GeometryTransform {
  unit: LengthUnit;
  scale: number;
  rotate_x: number;
  rotate_y: number;
  rotate_z: number;
  up_axis: UpAxis;
  center_xy: boolean;
  ground_align: boolean;
  invert_normals: boolean;
  simplify_target_faces: number | null;
}

export const DEFAULT_GEOMETRY_TRANSFORM: GeometryTransform = {
  unit: 'mm',
  scale: 1,
  rotate_x: 0,
  rotate_y: 0,
  rotate_z: 0,
  up_axis: 'z',
  center_xy: false,
  ground_align: false,
  invert_normals: false,
  simplify_target_faces: null,
};

export interface VoxelizeResult {
  geometry_id: string;
  grid_id: string;
  grid_path: string;
  grid_shape: number[];
  total_cells: number;
  solid_cells: number;
  fluid_cells: number;
}
